server/src-lib/Hasura/Server/Cors.hs

```diff
@@ -0,0 +1,154 @@
+{-# LANGUAGE DeriveAnyClass #-}
+
+-- | CORS (Cross Origin Resource Sharing) related configuration
+
+module Hasura.Server.Cors
+  ( CorsConfig (..)
+  , CorsPolicy (..)
+  , parseOrigin
+  , readCorsDomains
+  , mkDefaultCorsPolicy
+  , isCorsDisabled
+  , Domains (..)
+  ) where
+
+import           Hasura.Prelude
+import           Hasura.Server.Utils  (fmapL)
+
+import           Control.Applicative  (optional)
+
+import qualified Data.Aeson           as J
+import qualified Data.Aeson.Casing    as J
+import qualified Data.Aeson.TH        as J
+import qualified Data.Attoparsec.Text as AT
+import qualified Data.HashSet         as Set
+import qualified Data.Text            as T
+
+
+data DomainParts =
+  DomainParts
+  { wdScheme :: !Text
+  , wdHost   :: !Text -- the hostname part (without the *.)
+  , wdPort   :: !(Maybe Int)
+  } deriving (Show, Eq, Generic, Hashable)
+
+$(J.deriveToJSON (J.aesonDrop 2 J.snakeCase) ''DomainParts)
+
+data Domains
+  = Domains
+  { dmFqdns     :: !(Set.HashSet Text)
+  , dmWildcards :: !(Set.HashSet DomainParts)
+  } deriving (Show, Eq)
+
+$(J.deriveToJSON (J.aesonDrop 2 J.snakeCase) ''Domains)
+
+data CorsConfig
+  = CCAllowAll
+  | CCAllowedOrigins Domains
+  | CCDisabled
+  deriving (Show, Eq)
+
+instance J.ToJSON CorsConfig where
+  toJSON c = case c of
+    CCDisabled         -> toJ True J.Null
+    CCAllowAll         -> toJ False (J.String "*")
+    CCAllowedOrigins d -> toJ False $ J.toJSON d
+    where
+      toJ dis origs =
+        J.object [ "disabled" J..= dis
+                 , "allowed_origins" J..= origs
+                 ]
+
+isCorsDisabled :: CorsConfig -> Bool
+isCorsDisabled = \case
+  CCDisabled -> True
+  _          -> False
+
+readCorsDomains :: String -> Either String CorsConfig
+readCorsDomains str
+  | str == "*"               = pure CCAllowAll
+  | str == "" || str == ","  = Left "invalid domain"
```

Comment:
This check is not exhaustive (what if the string is " " or ",,"). Why not let the domain parsing logic handle empty strings post the split on `,`?

Reply:
Yes, this check is not necessary. It was remnant of older regex-based parsing logic. Removed it.